pex/pex.py

```diff
@@ -617,10 +617,47 @@ def execute_script(self, script_name):
         dist_script = get_script_from_distributions(script_name, dists)
         if not dist_script:
             return "Could not find script {!r} in pex!".format(script_name)
-        TRACER.log("Found script {!r} in {!r}.".format(script_name, dist))
-        return self.execute_content(
-            dist_script.path, dist_script.read_contents(), argv0=script_name
-        )
+        TRACER.log("Found script {!r} in {!r}.".format(script_name, dist_script.dist))
+
+        contents = dist_script.read_contents()
+        shebang = contents.split("\n", 1)[0]
+        try:
+            interpreter = PythonInterpreter.from_shebang(shebang)
+            if interpreter and interpreter.version_string != self.interpreter.version_string:
+                TRACER.log(
+                    "Python script shebang resolved to {shebang_python}, but we will execute it with {pex_python}".format(
+                        shebang_python=interpreter.version_string,
+                        pex_python=self.interpreter.version_string,
+                    ),
+                    V=2,
+                )
+        except PythonInterpreter.InterpreterNotFound as e:
+            TRACER.log(
+                "Python script shebang indicated unknown interpreter: {shebang!r}: {e}, will execute it with {version}.".format(
+                    shebang=shebang, e=e, version=self.interpreter.version_string
+                ),
+                V=2,
+            )
+            interpreter = self.interpreter
+
+        if interpreter:
+            TRACER.log(
+                "Executing python script with {}".format(self.interpreter.version_string), V=2
+            )
+            return self.execute_content(dist_script.path, contents, argv0=script_name)
+        else:
+            TRACER.log("Executing non-python script using shebang: {}".format(shebang), V=2)
+            return self.execute_external(dist_script.path)
+
+    def execute_external(self, cmd):
+        # type: (str) -> Any
+        try:
+            # execute as a shell command, inheriting stdout and stderr
+            proc = Executor.open_process([cmd] + sys.argv[:1], shell=True)
```

Comment:
This happens to work, but perhaps it bothered you why it worked? PEX files are zips after all.

Reply:
No, not really. Which shows that I haven't fully grasped what's going on :P

Reply:
Ok. The short of it is that all the distributions needed to run a PEX are extracted from the PEX file to disk (under ~/.pex/installed_wheels) and the locations of each are added to the `sys.path`. See the last 3 lines here:
https://github.com/pantsbuild/pex/blob/78c7b01d11eda2a70d1c0fb677c52b4f36a423f6/pex/environment.py#L329-L345

This is mainly to handle native extensions which can't be loaded from inside a zip like `.py` files can.